src/lightning/fabric/utilities/optimizer.py

```diff
@@ -31,4 +52,104 @@ def _optimizers_to_device(optimizers: Iterable[Optimizer], device: _DEVICE) -> N
 def _optimizer_to_device(optimizer: Optimizer, device: _DEVICE) -> None:
     """Moves the state of a single optimizer to the device."""
     for p, v in optimizer.state.items():
-        optimizer.state[p] = apply_to_collection(v, Tensor, move_data_to_device, device, allow_frozen=True)
+        optimizer.state[p] = apply_to_collection(v, torch.Tensor, move_data_to_device, device, allow_frozen=True)
+
+
+def _no_op() -> None:
+    pass
+
+
+@contextlib.contextmanager
+def _apply_optimizers_during_fsdp_backward(
+    optimizers: Iterable[Optimizer],
+    module_wrapper: "_FabricModule",
+) -> Generator[None, None, None]:
+    """Call `Optimizer.step` as gradients become available.
+
+    NOTE: This is an EXPERIMENTAL utility and exploits behavior which is not
+          part of the FSDP public API. Use at your own risk.
+
+    By moving optimizer step invocation into the backward call we can free
+    gradients earlier and reduce peak memory.
+    """
+    assert SUPPORTS_OPTIMIZER_IN_FSDP_BACKWARD
+    apply_lock = threading.Lock()
+
+    from lightning.fabric.wrappers import _FabricModule
+
+    assert isinstance(module_wrapper, _FabricModule)
+    param_handles = _get_fsdp_handles(module := module_wrapper._forward_module)
```

Comment:
Could we have a protected pytorch-only implementation that doesn't rely on fabric and a public (but experimental) implementation that does? For example:


```py
def _apply_optimizers_during_fsdp_backward():
    ...

def fsdp_overlap_step_with_backward(fabric_module, optimizers):
    return _apply_optimizers_during_fsdp_backward(optimizers, fabric_module._forward_module)

```

Reply:
Great point!